### ISSUE TITLE
[Feature Request] Operations between TensorDicts

### ISSUE THREAD

**Issue by Daxo32** [pytorch-labs/tensordict#181]
## Motivation

It would be really useful if there would be the possibility to perform operations such as _sum, multiplication_ etc... between two TensorDicts ( key-wise). Is this a feature that you have in mind to introduce in future?

Right now I noticed that is not possible.

## Checklist

- [ X] I have checked that there is no similar issue in the repo (**required**)

**Comment by vmoens** [comment]
Thanks for the question!
Yes that would be cool but what I don't want is to have algebraic operations implemented per se, bc they would only be used by a small subset of tensordicts and they would have undefined behaviours in many cases: this is why we don't have a dtype etc. 
E.g. what would it mean to do `tensordict + 1` if you have a mask in your tensordict? Should you ignore it? Should it be all set to True? Should it become a float?

## What you can already do

Regarding algebraic operations, you can already do `tensordict.apply(lambda x: x+1)` or similar. But I agree with you that we can't do that with 2 tensordicts / tensorclasses.

## What we could implement

I could see this:

```python
import tensordict

td3 = tensordict.apply(lambda x, y: x+y, td1, td2)
```
where td1 and td2 have a matching structure.
We could also do
```python
import tensordict

td4 = tensordict.apply(lambda x, y, z: x+y * z, td1, td2, td3)
```
etc.

That should also cover tensorclass.

Wdyt?

**Comment by Daxo32** [comment]
Hi @vmoens , thanks for the quick reply.

I understand you reasoning but I still think that those operation should be available. I have some reasons for that, please correct me if I am missing something here.

_> "E.g. what would it mean to do tensordict + 1 if you have a mask in your tensordict?"_

Well, I would expect that the output would match the exact behaviour of a tensor, if a tensor is masked (lets say a tensor of bools) then it may not make sense to add 1, but it is still doable.

`k = torch.zeros((1,5), dtype=torch.bool)) `
`k + 1`

`//returns:`
`tensor([[1, 1, 1, 1, 1]])`
`
is not reasonable to expect the same from a tensordict? 

We should not make any assumption on the semantics of a tensordict, the content of a tensordict may contain images, a mask,  or random numbers, but at the end are just numbers.
The keys of a dict are just strings and they are representative **only**  for the developer (it is not like a class), that's why it should be up to the developer on which operations apply to its values (like happens with the tensors).

Let me know if this makes sense or if I am missing something.

Thanks

**Comment by vmoens** [comment]
I see your point.
For these use cases you can do `td.apply(lambda x: x+1)` and it returns a tensordict with all the keys updated. Do you think that implementing `__add__` and others would bring a significant value to the lib? 
For instance, we don't implement grad but a nice feature of the lib is that you can store parameters in a tensordict and get all the grads just by doing 

`td.apply(lambda x: x.grad)`

One thing we could do is implement a fallback for whenever a method does not exist, ie if there is not add or matmul calling one of these falls back on calling it over all the tensors and wrapping the results in a tensordict.

Changing topic: what do you think of the apply method with another tensordict?
Should it be a method like 
`def apply_to(self, fun, *other: TensorDict)` or a function like suggested above?

**Comment by vmoens** [comment]
@Daxo32 

We talked about this and I don't think we'll support algebraic operations soon (operations across TDs is a different topic and I'll open a PR with it soon!)

Let me write down the various reasons here:

1. I think apply satisfies these needs already with little effort from developers, and any fallback would actually just call `apply` so the computational advantage of custom native implementations would be at most marginal.
2. A fallback like 
```python
def __getattr__(self, method):
    return lambda *args, **kwargs: self.apply(lambda x: x.method(*args, **kwargs))
```
would work in theory BUT here's the catch. We have not implemented everything yet.
TensorDict is designed to work only across batch dimensions (read: leading dimensions). Now take one func we don't cover yet (say, `Tensor.unique(dim=-1)`). If we do an implicit fallback, `TensorDict.unique(-1)` collect the unique over the last dimension of the "feature" dimensions, not the last dim of the batch. So we'd be buying a ticket for multiple new issues every time a function is not covered, and not because a missing attribute error has been raised but much worse, because the behaviour is not the one intended (ie can cause silent bugs).
3. Along the same line, although unitary ops can sometimes be meaningful, what about `td1 @ td2`? Should we do it across batch dims? I guess many wouldn't expect that. What about feature dims then? Well in that case we'd need to document 
> TensorDict methods work across batch dims only except for `matmul`, `dot` product etc
which would be a bit weird IMO. We would like for TensorDict to stay "intuitive", ie avoid any commitment to behaviours that require long explanations (although this may not be systematically avoidable, as in this case :) )
4. Having unitary ops only would mean that you could potentially have a code like this:
```python
def affine(td1, td2, td3):
    return apply_to(lambda x, y: x @ y, td1, td2) + td3
```
which you can be sure will cause people to come again and ask for matmul to be added to `TensorDict`.

Does that make sense to you?
Again, happy to read your thoughts on this, and see if there's a case for algebraic ops.

**Comment by Daxo32** [comment]
Hello @vmoens , thank you very much ! 
I also thought about this once again and I completely follow your idea and understand you reasoning. The _apply_ method is a good compromise to handle algebraic operations, I am looking forward for the TDs operations! I will follow this repo for sure, it is really interesting.

Thanks :)

**Comment by vmoens** [comment]
Hello @Daxo32 
We have recently merged #722 which positively solves this issue! Hope you'll like the feature.
